rust_src/src/fns.rs

```diff
@@ -362,4 +362,32 @@ pub fn load_average(use_floats: bool) -> Vec<LispNumber> {
         .collect()
 }
 
+/// Return a copy of ALIST.
+/// This is an alist which represents the same mapping from objects to objects,
+/// but does not share the alist structure with ALIST.
+/// The objects mapped (cars and cdrs of elements of the alist)
+/// are shared, however.
+/// Elements of ALIST that are not conses are also shared.
+#[lisp_fn]
+pub fn copy_alist(mut alist: LispObject) -> LispObject {
+    if alist.is_nil() {
+        return alist;
+    }
+
+    let new_alist = unsafe { lisp_concat(1, &mut alist, Lisp_Type::Lisp_Cons, false) };
+
+    let mut tem = new_alist;
+    while tem.is_not_nil() {
+        debug_assert!(tem.is_cons());
+        let t = tem.as_cons().unwrap();
+        let car = t.car();
+
+        if let Some(c) = car.as_cons() {
+            setcar(t, LispObject::cons(c.car(), c.cdr()));
```

Comment:
So `t` is the memory location pointed at by the `LispObject` `tem`? This is getting into confusing object identity territory.

Reply:
So `t` is a cons cell composed of `car` and `cdr`, and `car is a cons cell composed of `c.car()` and `c.cdr()`. Then we take those last two items, and recompose them into a new cons cell, then make that new cons cell the new car of `t`. Is that right?

Reply:
https://github.com/Wilfred/remacs/pull/1262 should make that call to `LispObject::cons` unnecessary.

Reply:
The version I recommend above is much simpler.